### PR TITLE
Imp: destroy invoker smoothly

### PR DESCRIPTION
**Pull request by AlexStocks**
**What this PR does**:
fix bug: protocol/dubbo/dubbo_invoker.go:DubboInvoker.reqNum should be a atomic number

### CODE REVIEW

File: protocol/dubbo/dubbo_invoker.go [hunk 1]
```diff
@@ -62,7 +62,8 @@ type DubboInvoker struct {
 	// timeout for service(interface) level.
 	timeout time.Duration
 	// Used to record the number of requests. -1 represent this DubboInvoker is destroyed
-	reqNum int64
+	reqNum uatomic.Int64
+	closed  uatomic.Bool
```
Comment (gaoxinge) [inline protocol/dubbo/dubbo_invoker.go:66]: `go fmt`.

**Review by wenxuwan — COMMENTED** [review]
这里使用原子操作和前面的没什么区别，还是建议使用读写锁来做

**Comment by gaoxinge** [comment]
I think some specific interleaving execution of `Invoke` and `Destroy` may lead to panic:

|Invoke|Destroy|Description|
|-------|---------|------------|
|di.closed.Load()||return false because destroy has not been called|
||di.closed.Store(true)||
||di.reqNum.Load() == 0|return true because no reqNum has been added in invoke|
||di.reqNum.Add(-1)||
||di.client = nil|set di.client to nil|
|di.reqNum.Add(1)|||
|di.client.Request||may panic because di.client is nil||
|di.reqNum.Add(-1)|||

File: protocol/dubbo/dubbo_invoker.go [hunk 2]
```diff
@@ -123,18 +138,22 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
...
 	rest := &protocol.RPCResult{}
 	timeout := di.getTimeout(inv)
-	if async {
-		if callBack, ok := inv.CallBack().(func(response common.CallbackResponse)); ok {
-			//result.Err = di.client.AsyncCall(NewRequest(url.Location, url, inv.MethodName(), inv.Arguments(), inv.Attachments()), callBack, response)
-			result.Err = di.client.AsyncRequest(&invocation, url, timeout, callBack, rest)
-		} else {
-			result.Err = di.client.Send(&invocation, url, timeout)
-		}
+	client := di.getClient()
```
Comment (zouyx) [inline protocol/dubbo/dubbo_invoker.go:141]: di.BaseInvoker.IsAvailable() 时，同时返回 client 是否可以减少一次再加锁的机会？

File: protocol/dubbo/dubbo_invoker.go [hunk 3]
```diff
@@ -162,27 +181,38 @@ func (di *DubboInvoker) getTimeout(invocation *invocation_impl.RPCInvocation) ti
...
 func (di *DubboInvoker) Destroy() {
 	di.quitOnce.Do(func() {
+		di.BaseInvoker.Stop()
+		var times int64
 		for {
-			if di.reqNum == 0 {
-				di.reqNum = -1
-				logger.Infof("dubboInvoker is destroyed,url:{%s}", di.GetUrl().Key())
+			times = di.BaseInvoker.InvokeTimes()
+			if times == 0 {
```
Comment (zouyx) [inline protocol/dubbo/dubbo_invoker.go:199]: 判断标准应该是 client 还是 times ?
我觉得还是像以前一样都兼顾一下比较好

File: protocol/grpc/grpc_invoker.go [hunk 4]
```diff
@@ -83,21 +127,47 @@ func (gi *GrpcInvoker) Invoke(ctx context.Context, invocation protocol.Invocatio
...
 	gi.quitOnce.Do(func() {
-		gi.BaseInvoker.Destroy()
-
-		if gi.client != nil {
-			_ = gi.client.Close()
+		gi.BaseInvoker.Stop()
+		var times int64
+		for {
+			times = gi.BaseInvoker.InvokeTimes()
+			if times == 0 {
```
Comment (zouyx) [inline protocol/grpc/grpc_invoker.go:155]: 同上

File: protocol/dubbo/dubbo_invoker.go [hunk 5]
```diff
@@ -87,15 +93,24 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
 		err    error
 		result protocol.RPCResult
 	)
-	if atomic.LoadInt64(&di.reqNum) < 0 {
+	if !di.BaseInvoker.IsAvailable() {
```
Comment (gaoxinge) [inline protocol/dubbo/dubbo_invoker.go:96]: This check is duplicated to below. What is it intended for?

File: protocol/dubbo/dubbo_invoker.go [hunk 6]
```diff
@@ -162,27 +181,38 @@ func (di *DubboInvoker) getTimeout(invocation *invocation_impl.RPCInvocation) ti
...
 	di.quitOnce.Do(func() {
+		di.BaseInvoker.Stop()
+		var times int64
 		for {
-			if di.reqNum == 0 {
-				di.reqNum = -1
-				logger.Infof("dubboInvoker is destroyed,url:{%s}", di.GetUrl().Key())
+			times = di.BaseInvoker.InvokeTimes()
+			if times == 0 {
+				di.BaseInvoker.AddInvokerTimes(-1)
```
Comment (gaoxinge) [inline protocol/dubbo/dubbo_invoker.go:200]: Why set invoke time to -1?

File: protocol/dubbo/dubbo_invoker.go [hunk 7]
```diff
@@ -162,27 +181,38 @@ func (di *DubboInvoker) getTimeout(invocation *invocation_impl.RPCInvocation) ti
...
-				if di.client != nil {
-					di.client.Close()
-					di.client = nil
+				client := di.getClient()
+				if client != nil {
+					client.Close()
+					di.setClient(nil)
 				}
 				break
+			} else if times < 0 {
```
Comment (wenxuwan) [inline protocol/dubbo/dubbo_invoker.go:209]: This check not needed

File: protocol/dubbo/dubbo_invoker.go (hunk 2 above)
Comment (wenxuwan) [inline protocol/dubbo/dubbo_invoker.go:141]: getClient返回的是client指针，下面关闭的时候，同样会把这个client赋值为nil

File: protocol/dubbo/dubbo_invoker.go (hunk 5 above)
Comment (flycash) [reply protocol/dubbo/dubbo_invoker.go:96]: It's double-check pattern, I think.

File: protocol/dubbo/dubbo_invoker.go (hunk 6 above)
Comment (flycash) [reply protocol/dubbo/dubbo_invoker.go:200]: -1 indicates `BaseInvoker` is unavailable

File: protocol/grpc/grpc_invoker.go [hunk 8]
```diff
@@ -83,21 +127,47 @@ func (gi *GrpcInvoker) Invoke(ctx context.Context, invocation protocol.Invocatio
...
+				gi.BaseInvoker.AddInvokerTimes(-1)
+				logger.Infof("grpcInvoker is destroyed, url:{%s}", gi.GetUrl().Key())
+				gi.BaseInvoker.Destroy()
+				client := gi.getClient()
+				if client != nil {
+					client.Close()
+					gi.setClient(nil)
```
Comment (flycash) [inline protocol/grpc/grpc_invoker.go:162]: I am not sure whether exchange those two lines' order:
```
gi.setClient(nil)
client.Close()
```
It looks more safe.

Comment (AlexStocks) [reply protocol/grpc/grpc_invoker.go:162]: good advice.

**Comment by wenxuwan** [comment]
这边可以用给一个原子操作代表invoker是不是可用状态，然后加一个读写锁，invoke的时候先用原子操作判断是不是可用的，如果可用加读锁，然后再通过原子操作判断一次，如果还可用，调用下面的call,完成后释放读锁。destroy的时候直接把invoker标记为不可用状态，然后阻塞获取写锁，获取成功后然后关闭invoker，释放client就可以了。

File: protocol/dubbo/dubbo_invoker.go (hunk 5 above)
Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go:96]: yes.

File: protocol/dubbo/dubbo_invoker.go (hunk 7 above)
Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go:209]: 万一发生什么幺蛾子呢？防御性编程，这里面有日志的。

File: protocol/grpc/grpc_invoker.go (hunk 8 above)
Comment (AlexStocks) [reply protocol/grpc/grpc_invoker.go]: have fixed it. pls check again.

**Review by flycash — APPROVED** [review]
LGTM

**Comment by AlexStocks** [comment]
@flycash pls check it again. I have reconstructed this pr as @wenxuwan's suggestion.

File: protocol/dubbo/dubbo_invoker.go [hunk 9]
```diff
@@ -101,8 +101,12 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
 		return &result
 	}
 
-	di.AddInvokerTimes(1)
-	defer di.AddInvokerTimes(-1)
+	client := di.getClient()
+	if client == nil {
+		result.Err = protocol.ErrClientClosed
+		logger.Debugf("result.Err: %v", result.Err)
```
Comment (flycash) [inline protocol/dubbo/dubbo_invoker.go:107]: How about using Warn? I think this only happens in some abnormal cases.

**Comment by flycash** [comment]
> @flycash pls check it again. I have reconstructed this pr as @wenxuwan's suggestion.

LGTM

File: protocol/dubbo/dubbo_invoker.go [hunk 10]
```diff
@@ -87,15 +93,28 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
...
+	if !di.BaseInvoker.IsAvailable() {
 		// Generally, the case will not happen, because the invoker has been removed
 		// from the invoker list before destroy,so no new request will enter the destroyed invoker
 		logger.Warnf("this dubboInvoker is destroyed")
-		result.Err = ErrDestroyedInvoker
+		result.Err = protocol.ErrDestroyedInvoker
+		return &result
+	}
+
+	client := di.getClient()
```
Comment (wenxuwan) [inline protocol/dubbo/dubbo_invoker.go:104]: 这个地方个人感觉完全可以把锁提出来

di.clientGuard.RLock()
client := di.client
defer di.clientGuard.RUnlock()

这样可以保证有请求的时候，不会说获取到client之后，到后面的client.call的时候，报链接关闭的错误。

Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go:104]: 关键是，这样锁粒度太大了。可能还是原来的方案好点，顶多在 destroy 里面阻塞下。你再想想

Comment (wenxuwan) [reply protocol/dubbo/dubbo_invoker.go:104]: 力度不大的，destroy本身就只会调用一次，invoke的时候拿的是读锁，destroy的时候才会发生互斥。所以你拿出来不拿出来都是在destroy那边阻塞一下

Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go]: done

File: protocol/dubbo/dubbo_invoker.go [hunk 11]
```diff
@@ -87,15 +93,31 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
...
 		logger.Warnf("this dubboInvoker is destroyed")
-		result.Err = ErrDestroyedInvoker
+		result.Err = protocol.ErrDestroyedInvoker
+		return &result
+	}
+
+	di.clientGuard.RLock()
+	defer di.clientGuard.RUnlock()
+
+	client := di.client
```
Comment (wenxuwan) [inline protocol/dubbo/dubbo_invoker.go:107]: here we use Rlock to protect client, so we can use di.client directly, new variable client is not needed

File: protocol/dubbo/dubbo_invoker.go [hunk 12]
```diff
@@ -87,15 +93,31 @@ func (di *DubboInvoker) Invoke(ctx context.Context, invocation protocol.Invocati
...
+	if client == nil {
+		result.Err = protocol.ErrClientClosed
```
Comment (wenxuwan) [inline protocol/dubbo/dubbo_invoker.go:109]: may be client should never be nil with di.clientGuard.RLock() ?

File: protocol/dubbo/dubbo_invoker.go (hunk 11 above)
Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go:107]: can agree with u more.

**Review by gaoxinge — APPROVED** [review]

File: protocol/dubbo/dubbo_invoker.go (hunk 9 above)
Comment (AlexStocks) [reply protocol/dubbo/dubbo_invoker.go:109]: good suggestion.

**Review by wenxuwan — APPROVED** [review]

**Review by zouyx — APPROVED** [review]

**Pull request merged**